src/table/iterator.rs

```diff
@@ -0,0 +1,170 @@
+use super::builder::{Header, HEADER_SIZE};
+use super::{Block, TableInner};
+use crate::util::{self, KeyComparator, COMPARATOR};
+use crate::value::Value;
+use bytes::Bytes;
+use std::sync::Arc;
+
+#[derive(Clone, Debug)]
+pub enum IteratorError {
+    NoError,
+    EOF,
+    // TODO: As we need to clone Error from block iterator to table iterator,
+    // we had to save `crate::Error` as String. In the future, we could let all
+    // seek-related function to return a `Result<()>` instead of saving an
+    // error inside struct.
+    Error(String),
+}
+
+impl IteratorError {
+    pub fn is_err(&self) -> bool {
+        match self {
+            IteratorError::NoError => false,
+            _ => true,
+        }
+    }
+
+    pub fn is_eof(&self) -> bool {
+        match self {
+            IteratorError::EOF => true,
+            _ => false,
+        }
+    }
+}
+
+enum SeekPos {
+    Origin,
+    Current,
+}
+
+/// Block iterator iterates on an SST block
+// TODO: support custom comparator
+struct BlockIterator {
+    idx: isize,
+    base_key: Bytes,
```

Comment:
How about using `BytesMut` instead? So you can clone the diff key without allocating frequently.

Reply:
I've changed `key` to `BytesMut`, and now we could construct key without allocating new space.